lib/Decl.cpp

```diff
@@ -188,4 +195,131 @@ llvm::Value *FunctionDecl::CallFromLiftedBlock(
+  }
+  return json;
+}
+
+// Serialize a ParameterDecl to JSON
+llvm::json::Object ParameterDecl::SerializeToJSON(const llvm::DataLayout &dl) {
+  // Get the serialization for the ValueDecl
+  ValueDecl *val_decl_ptr = dynamic_cast<ValueDecl *>(this);
+  llvm::json::Object param_json = val_decl_ptr->SerializeToJSON(dl);
+
+  // Insert "name"
+  param_json.insert(
+      llvm::json::Object::KV{llvm::json::ObjectKey("name"), this->name});
+
+  return param_json;
+}
+
+// Serialize a ValueDecl to JSON
+llvm::json::Object ValueDecl::SerializeToJSON(const llvm::DataLayout &dl) {
+  llvm::json::Object value_json;
+
+  if (this->reg) {
+    // The value is in a register
+    value_json.insert(llvm::json::Object::KV{llvm::json::ObjectKey("register"),
+                                             this->reg->name});
+  } else if (this->mem_reg) {
+    // The value is in memory
+    llvm::json::Object memory_json;
+    memory_json.insert(llvm::json::Object::KV{llvm::json::ObjectKey("register"),
+                                              this->mem_reg->name});
+    memory_json.insert(llvm::json::Object::KV{llvm::json::ObjectKey("offset"),
+                                              this->mem_offset});
+
+    // Wrap the memory_json structure in a memory block
+    value_json.insert(
+        llvm::json::Object::KV{llvm::json::ObjectKey("memory"),
+                               llvm::json::Value(std::move(memory_json))});
+  } else {
+    LOG(ERROR) << "Trying to serialize a value that has not been allocated";
+    exit(1);
+  }
+
+  value_json.insert(llvm::json::Object::KV{llvm::json::ObjectKey("type"),
+                                           TranslateType(*this->type, dl)});
+
+  return value_json;
+}
+
+// Create a Function Declaration from llvm::Function
+FunctionDecl FunctionDecl::Create(const llvm::Function &func,
+                                  const llvm::Module &mdl,
+                                  const remill::Arch::ArchPtr &arch) {
+  const llvm::DataLayout& dl = mdl.getDataLayout();
+
+  FunctionDecl decl;
+  decl.type = func.getFunctionType();
+  decl.name = func.getName().data();
+  decl.demangled_name = llvm::demangle(func.getName().data());
+  decl.dl = &dl;
+
+  // If the function calling convention is not the default llvm::CallingConv::C
+  // then use it. Otherwise, get the CallingConvention from the remill::Arch
+  std::unique_ptr<CallingConvention> cc;
+  llvm::CallingConv::ID cc_id = func.getCallingConv();
+  if (cc_id != llvm::CallingConv::C) {
+    // The value is not default so use it
+    cc = CallingConvention::CreateCCFromCCID(cc_id, arch.get());
+  } else {
+    cc = CallingConvention::CreateCCFromArch(arch.get());
+  }
+
+  cc->AllocateSignature(decl, func);
+
+  // TODO(aty): for a better and more comprehensive serialization
+  // decl->address =
+  // decl->return_address =
+  // decl->return_stack_pointer =
+  // decl->return_stack_pointer_offset =
+  // decl->is_noreturn =
+  // decl->is_variadic =
```

Comment:
`is_noreturn` and `is_variadic` are pretty easy to handle. You can do: `func.hasFnAttr(llvm::Attribute::NoReturn)` and probably `func.isVariadic()`, respectively.